compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs

```diff
@@ -0,0 +1,162 @@
+        T.concat ["enum constructor ", dot m, ":", dot t, ".", v]
+    NField (ModuleName m) (TypeConName t) (FieldName f) ->
+        T.concat ["field ", dot m, ":", dot t, ".", f]
+    NChoice (ModuleName m) (TypeConName t) (ChoiceName c) ->
+        T.concat ["choice ", dot m, ":", dot t, ".", c]
+  where
+    dot = T.intercalate "."
+
+-- | Asks whether a name collision is permitted. According to the
+-- LF Spec, a name collision is only permitted when it occurs
+-- between a record type and a variant constructor defined in
+-- the same module.
+nameCollisionPermitted :: Name -> Name -> Bool
+nameCollisionPermitted a b =
+    case (a,b) of
+        (NRecordType m1 _, NVariantCon m2 _ _) -> m1 == m2
+        (NVariantCon m1 _ _, NRecordType m2 _) -> m1 == m2
+        _ -> False
+
+-- | Asks whether a name collision is forbidden.
+nameCollisionForbidden :: Name -> Name -> Bool
+nameCollisionForbidden a b = not (nameCollisionPermitted a b)
+
+-- | Fully resolved name within a package. We don't use
+-- Qualified from DA.Daml.LF.Ast because that hides collisions
+-- between module names and type names. This should only be
+-- constructed lower case in order to have case-insensitivity.
+--
+-- This corresponds to the following section of the LF spec:
+-- https://github.com/digital-asset/daml/blob/master/daml-lf/spec/daml-lf-1.rst#fully-resolved-name
+newtype FRName = FRName [T.Text]
+    deriving (Eq, Ord)
+
+-- | Turn a name into a fully resolved name.
+fullyResolve :: Name -> FRName
+fullyResolve = FRName . map T.toLower . \case
+    NModule (ModuleName m) ->
+        m
+    NRecordType (ModuleName m) (TypeConName t) ->
+        m ++ t
+    NVariantType (ModuleName m) (TypeConName t) ->
+        m ++ t
+    NEnumType (ModuleName m) (TypeConName t) ->
+        m ++ t
+    NVariantCon (ModuleName m) (TypeConName t) (VariantConName v) ->
+        m ++ t ++ [v]
+    NEnumCon (ModuleName m) (TypeConName t) (VariantConName v) ->
+        m ++ t ++ [v]
+    NField (ModuleName m) (TypeConName t) (FieldName f) ->
+        m ++ t ++ [f]
+    NChoice (ModuleName m) (TypeConName t) (ChoiceName c) ->
+        m ++ t ++ [c]
+
+-- | State of the name collision checker. This is a
+-- map from fully resolved names within a package to their
+-- original names. We update this map as we go along.
+newtype NCState = NCState (M.Map FRName [Name])
+
+-- | Initial name collision checker state.
+initialState :: NCState
+initialState = NCState M.empty
+
+-- | Try to add a name to the NCState. Returns Error only
+-- if the name results in a forbidden name collision.
+addName :: Name -> NCState -> Either Error NCState
+addName name (NCState nameMap) = do
+    let frName = fullyResolve name
+        oldNames = fromMaybe [] (M.lookup frName nameMap)
+        badNames = filter (nameCollisionForbidden name) oldNames
+    if null badNames then do
+        Right . NCState $ M.insert frName (oldNames ++ [name]) nameMap
```

Comment:
Do we care about the order here? Otherwise, prepending rather than appending would make me more comfortable even though it looks like these lists can’t get long :slightly_smiling_face:

Reply:
Ah, yeah, it probably doesn't matter. I'm using `(:)` now :)